core/jobs/batch_jobs/manual_voice_artist_names_job.py

```diff
@@ -0,0 +1,504 @@
+        new_voiceover_mapping: Dict[
+            str, Dict[str, voiceover_models.VoiceoverDict]],
+        old_voiceover_mapping: Dict[
+            str, Dict[str, voiceover_models.VoiceoverDict]],
+        voice_artist_id_to_metadata_mapping: Dict[
+            str, voiceover_models.VoiceoversAndContentsMappingType],
+        exploration_id: str
+    ) -> Tuple[
+        Dict[str, Dict[str, voiceover_models.VoiceoverDict]],
+        Dict[str, voiceover_models.VoiceoversAndContentsMappingType]
+    ]:
+        """The method calculates the difference between the old and new values
+        of the recorded voiceover to isolate and retrieve only the updated
+        voiceover values within a given commit.
+
+        Args:
+            new_voiceover_mapping: dict(str, dict(str, VoiceoverDict)). A dict
+                representing new values of recorded voiceovers.
+            old_voiceover_mapping: dict(str, dict(str, VoiceoverDict)). A dict
+                representing old values of recorded voiceovers.
+            voice_artist_id_to_metadata_mapping:
+                dict(str, VoiceoversAndContentsMappingType). A dictionary
+                mapping voice artist IDs as keys and nested dicts as values.
+                Each nested dict refers to the voiceovers_and_contents_mapping
+                field in the VoiceArtistMetadataModel.
+            exploration_id: str. The ID of the given exploration.
+
+        Returns:
+            *. A 2-tuple consisting of:
+            - A dictionary representing the difference between the old and new
+            values of the recorded voiceover.
+            - Another dictionary representing the updated metadata mapping for
+            the voice artist.
+        """
+        voiceover_mapping_diff: Dict[
+            str, Dict[str, voiceover_models.VoiceoverDict]] = {}
+
+        for content_id, lang_code_to_voiceover_dict in (
+            new_voiceover_mapping.items()):
+
+            voiceover_mapping_diff[content_id] = {}
+
+            for lang_code, voiceover_dict in (
+                lang_code_to_voiceover_dict.items()):
+
+                if lang_code not in old_voiceover_mapping[content_id]:
+                    voiceover_mapping_diff[content_id][lang_code] = (
+                        voiceover_dict)
+                else:
+                    old_voiceover_dict = old_voiceover_mapping[
+                        content_id][lang_code]
+                    new_voiceover_dict = new_voiceover_mapping[
+                        content_id][lang_code]
+
+                    if old_voiceover_dict != new_voiceover_dict:
+                        voiceover_mapping_diff[content_id][lang_code] = (
+                            voiceover_dict)
+                        voice_artist_id_to_metadata_mapping = (
+                            CreateVoiceArtistMetadataModelsFromExplorationsJob.
+                            _delete_voice_artist_model_for_old_commits(
```

Comment:
I am a bit confused by what's going on here TBH. It seems like a lot of bookkeeping. I also don't quite know why we care about deletion ... surely a voiceover recorded by person X is still useful to listen to even though it's not being used any more?

Also I don't see anything that indicates you're processing these in increasing/decreasing order of exploration version?

One thought: have you considered going the other way, from latest to earliest revision? You can populate a dict of content-id-to-voiceover-artist for each language code, and avoid updating that dict for entries that have already been filled. Make sure to skip over any revert commits (e.g. if a commit reverts from version 192 to version 185 then when you reach it, skip backwards to version 184 and don't process versions 185-192). That way you don't need to do any deletions.

Reply:
If we keep the voiceovers and content IDs that are deleted (used earlier in some versions), this will cause confusion while migrating to new storage system because for each content there will be more than 1 voice artists information. And deciding who contributed to the latest voiceover will be tricky.

Since GetModels method of ndb_io give models in the sorted order, if we somehow reverse the list i.e., to iterate from version n to 1, we can face the following issues:
We need to maintain additional bookkeeping of exploration ID, content ID, language code and voice artist ID. 

Performing additional bookkeeping vs deleting old voice artist information if exists is simpler in terms understandibility.

Example:
![Screenshot from 2024-03-06 14-17-45](https://github.com/oppia/oppia/assets/57531197/ee862f26-b9b8-4136-b030-c903b59b6e50)

Reply:
@Nik-09 I don't follow -- my idea is that, when iterating backward, if you come to an artist for a "voiceover slot" which already has one assigned, you **don't** record the new artist at all. That way, each slot in the latest version of the exploration is mapped to the latest artist to contribute a voiceover for it, right?

Please let me know if this clarifies, or if there's something I'm missing here. Thanks.

Reply:
Updated the approach, PTAL